newton/_src/sim/builder.py

```diff
@@ -5890,98 +5913,152 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
                 )
             )
 
-            # Check if SDF generation was requested but we're on CPU
+            # Check if there are any shapes with hydroelastic collision enabled
+            has_hydroelastic_shapes = any(
+                (sflags & ShapeFlags.HYDROELASTIC) and (sflags & ShapeFlags.COLLIDE_SHAPES)
+                for sflags in self.shape_flags
+            )
+
+            # Validate that hydroelastic shapes have sdf_max_resolution or sdf_target_voxel_size set
+            for shape_idx, (sflags, sdf_max_resolution, sdf_target_voxel_size, shape_key) in enumerate(
+                zip(
+                    self.shape_flags,
+                    self.shape_sdf_max_resolution,
+                    self.shape_sdf_target_voxel_size,
+                    self.shape_key,
+                    strict=True,
+                )
+            ):
+                if (
+                    (sflags & ShapeFlags.HYDROELASTIC)
+                    and (sflags & ShapeFlags.COLLIDE_SHAPES)
+                    and sdf_max_resolution is None
+                    and sdf_target_voxel_size is None
+                ):
+                    raise ValueError(
```

Comment:
It would be better to have this validation inside `ShapeConfig.__post_init__()` because it will only be done at the place when the user actually passes these settings into the ModelBuilder (if the user manually overwrites the builder lists themselves we shouldn't need to accommodate this abuse of API). These kinds of loops over all shapes can really slow down the build step if you have a lot of environments.